onflow/flow-go
# [Awaiting Mainnet25] FLIP 204 - Smart-contract Defines target duration and end time for epochs

---

🚢 This PR is not downward-compatible. It should be merged after the last HCU of Mainnet24 and before the Mainnet25 network upgrade.

---

This PR merges the FLIP 204 feature branch, which implements [FLIP 204: Smart-Contract-Specified Epoch Switchover Timing](https://github.com/onflow/flips/blob/main/protocol/20231003-service-account-specified-epoch-switchover.md.md).

This branch contains changes from PRs:
- https://github.com/onflow/flow-go/pull/4987
- https://github.com/onflow/flow-go/pull/5038
- https://github.com/onflow/flow-go/pull/5023
- https://github.com/onflow/flow-go/pull/5027

And from resolving conflicts in the merge with `master`. 

<details>
  <summary>Description of conflict resolution that has already been reviewed</summary>
<ul>
<li><a href="https://github.com/onflow/flow-go/pull/5071/commits/2479faac8b2b301ad5575cd9676cbd37a74b488c">2479faac8b2b301ad5575cd9676cbd37a74b488c</a></li>
<li><a href="https://github.com/onflow/flow-go/pull/5071/commits/ef2a908862c51734072bfac227f51d1da13fbeba">ef2a908862c51734072bfac227f51d1da13fbeba</a></li>
<li>These conflicts are summarized by <a href="https://github.com/onflow/flow-go/pull/5071/files?file-filters%5B%5D=.mod&file-filters%5B%5D=.sum&file-filters%5B%5D=No+extension&show-deleted-files=true&show-viewed-files=true">this file-filtered view of the diff</a> .</li>
</details>

### How to Review
See https://github.com/onflow/flow-go/pull/5071#issuecomment-1922247785 for list of conflicting files and commits to review, following the sync on Feb 1 2024.

---

I was spending a lot of time addressing conflicts between the feature branch and `master` using a separate branch and PR (as `feature/*` branches are protected). This PR replaces https://github.com/onflow/flow-go/pull/5062  so I can resolve the merge conflicts directly without an extra layer of CI and review. This branch was created as follows:
- Delete local copy of branch `feature/flip-204-epoch-target-end-time`
- Pull `feature/flip-204-epoch-target-end-time` and `master` from GitHub, merge, address conflicts in `**/go.mod` files
- Switch onto a new branch
```sh
git checkout master
git branch -D feature/flip-204-epoch-target-end-time
git pull
git checkout feature/flip-204-epoch-target-end-time
git merge master
# Address conflicts in files:
# CONFLICT (content): Merge conflict in go.mod
# CONFLICT (content): Merge conflict in insecure/go.mod
# CONFLICT (content): Merge conflict in integration/go.mod
git checkout -b jordan/feat/flip204
git push -u origin
```
